### ISSUE TITLE
Support Base64 Encoding LargeBinary Field

### ISSUE THREAD

**Issue by johnthagen** [collerek/ormar#187]
**Is your feature request related to a problem? Please describe.**

In a `Model` such as:

```py
class BaseMessageModel(Model):
    class Meta:
        metadata = metadata
        database = database

    data: bytes = LargeBinary(max_length=1000)
```

One of the challenges that comes up is actually serializing this binary data effectively. If there are any UTF-8 incompatible bytes:

```
Traceback (most recent call last):
  File "venv/lib/python3.9/site-packages/uvicorn/protocols/http/httptools_impl.py", line 398, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "venv/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "venv/lib/python3.9/site-packages/fastapi/applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "venv/lib/python3.9/site-packages/starlette/applications.py", line 111, in __call__
    await self.middleware_stack(scope, receive, send)
  File "venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "venv/lib/python3.9/site-packages/starlette/middleware/cors.py", line 78, in __call__
    await self.app(scope, receive, send)
  File "venv/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "venv/lib/python3.9/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "venv/lib/python3.9/site-packages/starlette/routing.py", line 566, in __call__
    await route.handle(scope, receive, send)
  File "venv/lib/python3.9/site-packages/starlette/routing.py", line 376, in handle
    await self.app(scope, receive, send)
  File "venv/lib/python3.9/site-packages/fastapi/applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "venv/lib/python3.9/site-packages/starlette/applications.py", line 111, in __call__
    await self.middleware_stack(scope, receive, send)
  File "venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "venv/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "venv/lib/python3.9/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "venv/lib/python3.9/site-packages/starlette/routing.py", line 566, in __call__
    await route.handle(scope, receive, send)
  File "venv/lib/python3.9/site-packages/starlette/routing.py", line 227, in handle
    await self.app(scope, receive, send)
  File "venv/lib/python3.9/site-packages/starlette/routing.py", line 41, in app
    response = await func(request)
  File "venv/lib/python3.9/site-packages/fastapi/routing.py", line 209, in app
    response_data = await serialize_response(
  File "venv/lib/python3.9/site-packages/fastapi/routing.py", line 127, in serialize_response
    return jsonable_encoder(
  File "venv/lib/python3.9/site-packages/fastapi/encoders.py", line 104, in jsonable_encoder
    jsonable_encoder(
  File "venv/lib/python3.9/site-packages/fastapi/encoders.py", line 57, in jsonable_encoder
    return jsonable_encoder(
  File "venv/lib/python3.9/site-packages/fastapi/encoders.py", line 90, in jsonable_encoder
    encoded_value = jsonable_encoder(
  File "venv/lib/python3.9/site-packages/fastapi/encoders.py", line 127, in jsonable_encoder
    return ENCODERS_BY_TYPE[type(obj)](obj)
  File "pydantic/json.py", line 43, in pydantic.json.lambda
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xbe in position 1: invalid start byte
```

This makes sense since you can't send arbitrary bytes through a fundamentally text-based (UTF-8) JSON representation.

**Describe the solution you'd like**

```py
class BaseMessageModel(Model):
    class Meta:
        metadata = metadata
        database = database

    data: str = LargeBinary(max_length=1000, represent_as_base64=True)
```

And internally this would call something like:

```py
    @validator("data")
    def convert_to_base64_str(cls, v: bytes) -> str:
        return base64.b64encode(v).decode()
```

**Edit**: It looks like FastAPI doesn't use OpenAPI 3.1 yet (it is newly released), so probably the older 3.0 format should be used:

https://github.com/OAI/OpenAPI-Specification/blob/fbe62006211838a8bb7bf2433a1d15f1a5838a03/versions/3.0.1.md#considerations-for-file-uploads

```yaml
# content transferred with base64 encoding
schema:
  type: string
  format: base64
```

**Future Considerations**

In OpenAPI 3.1 (not yet supported by FastAPI): https://github.com/OAI/OpenAPI-Specification/issues/1547#issuecomment-595497718

```yaml
type: string
contentEncoding: base64
```

**Describe alternatives you've considered**

Ideally Pydantic would support this serialization upstream: https://github.com/samuelcolvin/pydantic/issues/692 but it's been that issue has been open for 2 years.

**Comment by collerek** [comment]
I will look into that, had same problems and initially did encode it, the question was if that should be:

*  configurable (probably yes - with flag at field definition)
*  default behavior (again, probably yes? I guess it's not common to send just byte representation of a string in this field and it's pretty common to get not utf-8 compliant bytes)

I will read the pydantic issue and see.

As I can overwrite validation in model anyway (that's what is happening with json fields -> ormar dumps/loads values for you so you can assign i.e. dict or list while pydantic.Json field accepts only str/bytes).

**Comment by johnthagen** [comment]
Yes, I was thinking through these seem issues. It seems like it should be configurable. Perhaps someone wants to use Ormar without FastAPI and they are sending their data in some manner other than JSON? I can't really come up with a real use case myself for wanting to do this, but it at least seems possible.

As for the default, given `ormar`'s users today are likely a vast majority FastAPI users, I would think that base64-encoding a binary field would be what most people would want to do. On the other hand, *not* base64 encoding is also the least surprising from a simple type-view perspective.

I think as long as the documentation clearly shows how to do both situations, then people shouldn't be surprised.

```py
class MyModel1(Model):
    data: bytes = LargeBinary(max_length=1000, represent_as_base64=False)

class MyModel1(Model):
    data: str = LargeBinary(max_length=1000, represent_as_base64=True)
```

**Comment by johnthagen** [comment]
From the DjangoRESTFramework ecosystem, one package uses the name `represent_in_base64`:

https://github.com/Hipo/drf-extra-fields/blob/252e681765645322113f69131e786664f815b919/drf_extra_fields/fields.py#L55

**Comment by johnthagen** [comment]
Hmm, I think we'll also want to test with `mypy` and the [`mypy` Pydantic plugin](https://pydantic-docs.helpmanual.io/mypy_plugin/#enabling-the-plugin):

```py
class MyBase64Model(Model):
    data: str = LargeBinary(max_length=1000)
```

```
$ mypy
...
error: Incompatible types in assignment (expression has type "LargeBinary", variable has type "str")
```

**Comment by johnthagen** [comment]
> It should be bytes not str as a type hint

I mean if/when we add support for representing as a base64 encoded `str`.

**Comment by collerek** [comment]
Yeah, sorry it's late here 🤣

**Comment by johnthagen** [comment]
```
b"\xbe".decode()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xbe in position 0: invalid start byte
```

**Comment by collerek** [comment]
OK I got entangled in this a little bit.
I kind of agree that with @samuelcolvin that you should be able to get both bytes and str from base64 encoded field.

Now my understanding is that it actually means that you have 3 scenarios:
1. As it was with initial `LargeBinary` field -> you pass and receive the bytes (utf-8 encoded) and bytes are stored in db
2. Base64 bytes  -> you pass and receive base64 encoded bytes, and store the same base64 encoded bytes in db
3. Base64 bytes as string -> you pass and receive base64 encoded string, but **in database you still store base64 encoded bytes**, not the string (so this one is basically to transfer data over json/api, but db should keep bytes)

Do I get it right, or I got confused along the way? 😄

**Comment by johnthagen** [comment]
> As it was with initial LargeBinary field -> you pass and receive the bytes (utf-8 encoded) and bytes are stored in db

I don't think that you can assume that `bytes` in the pure sense have any encoding. If there is a pure `bytes` -> `bytes` mode, then I think that the `bytes` should be stored purely as-is and returned as-is. This isn't super helpful for the FastAPI/JSON case because you can't automatically plug that into an OpenAPI endpoint and use it, but it's the most pure solution.

So, yeah how it currently works. The encoding is up to the user to deal with, and it won't work out of the box with FastAPI for all inputs.

> 2. Base64 bytes -> you pass and receive base64 encoded bytes, and store the same base64 encoded bytes in db

To me this sounds like the same as the first? If the user already base64 encoded the bytes, then it would be the same transparent `bytes` -> `bytes` storage and transmission. If the user does their base64 encoding themselves outside of `ormar`/Pydantic, then that is a case where 1. works. To me this seems like just a special, suggested documented case of 1. where `ormar` doesn't try to do anything special and it's just `bytes` in, `bytes` out, but don't go outside UTF-8 or this will fail.

There are other ways a user could choose to do `bytes` -> `bytes` as well, just as an example. They could encode all nibbles as ASCII hex number `344FF12`. This would be strictly less compressed than base64, but just used as an example.

> Base64 bytes as string -> you pass and receive base64 encoded string, but in database you still store base64 encoded bytes, not the string (so this one is basically to transfer data over json/api, but db should keep bytes)

Yes, this is case that will be most useful for FastAPI users. This allows you to interact with a `LargeBinary`/Blob SQL field without extra ceremony/boilerplate. You could imagine a system where binary data is being written to the database from some other process, and the web server just wants to be able to read and serialize that to API clients. By doing the automatic conversion of `bytes` -> Base64 encoded `str`, you can seamlessly expose that binary data in a way that is compatible with JSON.

___________

In summary, I think there are actually 2 cases (1 & 2, and 3), and it would be really useful to have that exposed to satisfy both common use cases.

**Comment by collerek** [comment]
I was thinking of supporting also auto string conversion (that's why it's 1 & 2) but maybe that does not make sense 🤔
For sure I want to allow passing str in case 3 so you can actually save the data by API call.

**Comment by johnthagen** [comment]
On 1 & 2 which side is the `str` conversion happening? I see `bytes` referenced on all ends.

I could very well be missing a use case as well, this is just my thoughts. Perhaps as more users pick up `ormar`, those other use cases will become more apparent.

**Comment by collerek** [comment]
Right now it's not happening, a possible use case would be to convert string to bytes on init/set and store it as bytes.
So by default it's bytes -> bytes, but if you provide a string it will be str -> convert -> bytes and later on fetch it will be bytes (so one way convert).

**Comment by collerek** [comment]
closed in #195

**Comment by johnthagen** [comment]
@collerek The new fixes to `LargeBinary` look great. I noticed a couple of items:

```py
class BaseMessageModel(Model):
    class Meta:
        metadata = metadata
        database = database

    data: str = LargeBinary(max_length=1000, represent_as_base64_str=True)
```

`data` has the correct type at runtime (`str`) but if using `mypy`, this will throw a type checking error at type checking time:

```
error: Incompatible types in assignment (expression has type "LargeBinary", variable has type "str")
```

Should `LargeBinary` support a type of `Union[bytes, str]`? Or perhaps another type `Base64LargeBinary` if you want to keep typing purity?

Secondly, the OpenAPI schema does not seem to be totally correct when using `represent_as_base64_str=True`:

<img width="328" alt="Screen Shot 2021-05-17 at 1 21 23 PM" src="https://user-images.githubusercontent.com/10340167/118530726-f86b2d80-b712-11eb-9a96-c8f9ada7ead3.png">

I think that this should be set as [`string($base64)` format](https://github.com/OAI/OpenAPI-Specification/blob/fbe62006211838a8bb7bf2433a1d15f1a5838a03/versions/3.0.1.md#considerations-for-file-uploads)? I'm not sure if `ormar` has control of that?

**Comment by collerek** [comment]
Ah shit completely forgotten about mypy check for str annotation 🤦 
Although it might be tricky as it uses the same field, will check typing overloads (they don't play well with kwargs).

The schema modifications are limited (you cannot change everything) but this one should be possible

**Comment by collerek** [comment]
That's what you get when one simple change requires major refactor of some aspects 🤣 
At least we gained some performance speed (and now it's more elegant, never liked overly complicated `__getattribute__`). 

Tracked in a new issue so now I don't miss it again ;)

**Comment by johnthagen** [comment]
> completely forgotten about mypy check for str annotation 

No worries! Thanks for taking a look at this!
